Applications/Spire/Source/Ui/Window.cpp

```diff
@@ -185,17 +194,56 @@ bool Window::nativeEvent(const QByteArray& eventType, void* message,
     }
   } else if(msg->message == WM_GETMINMAXINFO) {
     auto mmi = reinterpret_cast<MINMAXINFO*>(msg->lParam);
-    mmi->ptMaxTrackSize.x = maximumSize().width();
-    mmi->ptMaxTrackSize.y = maximumSize().height();
-    mmi->ptMinTrackSize.x = minimumSize().width();
-    mmi->ptMinTrackSize.y = minimumSize().height();
+    if(maximumSize() != QSize(QWIDGETSIZE_MAX, QWIDGETSIZE_MAX)) {
+      mmi->ptMaxTrackSize.x = maximumSize().width() +
+        2 * SYSTEM_BORDER_SIZE().width();
+      mmi->ptMaxTrackSize.y = maximumSize().height() +
+        SYSTEM_BORDER_SIZE().height();
+    } else {
+      mmi->ptMaxTrackSize.x = maximumSize().width();
+      mmi->ptMaxTrackSize.y = maximumSize().height();
+    }
+    mmi->ptMinTrackSize.x = minimumSize().width() +
+      2 * SYSTEM_BORDER_SIZE().width();
+    mmi->ptMinTrackSize.y = minimumSize().height() +
+      SYSTEM_BORDER_SIZE().height();
     return true;
   }
   return QWidget::nativeEvent(eventType, message, result);
 }
 
 void Window::resize_body(const QSize& size) {
-  resize({size.width(), size.height() + m_title_bar->height()});
+  if(!m_body) {
+    return;
+  }
+  resize(adjusted_window_size(size));
+}
+
+void Window::set_body(QWidget* body) {
+  set_body(body, body->size());
+}
+
+void Window::set_body(QWidget* body, const QSize& size) {
+  if(m_body) {
+    return;
+  }
+  m_body = body;
+  layout()->addWidget(m_body);
+  resize_body(size);
+}
+
+void Window::set_fixed_body(QWidget* body, const QSize& size) {
+  if(m_body) {
+    return;
+  }
+  set_window_attributes(false);
+  body->setFixedSize(size);
+  layout()->addWidget(body);
```

Comment:
I don't understand how the Window is resized after setting fixed body because I don't see the resize function is called.

Reply:
I think in this case the fixed size body in the layout is influencing the window size. I can't find where it would be documented, but a quick test in Qt Creator with a window that's initially 200x200 and a button that's 500x500 shows that the window will expand to fit the fixed size button, assuming its in the window's layout. In this same scenario, if the window has no layout then its shown at 200x200 pixels and most of the button overlaps (and is partly hidden by) the boundary of the window.

Reply:
I see. Thanks.